documentation/smg/10_build_release_process.md

```diff
@@ -0,0 +1,91 @@
+- all builds are created using the same build tools
+- users are using a known, traceable, validated release
+
+Disadvantages:
+
+- slower; have to wait for full build and test pipeline to execute
+
+### Releases
+
+The Jenkins build artifacts are are not accessible outside of STFC. End users will be directed to GitHub to access packaged releases.
+
+A 'deploy' pipeline will be created ([Horace #73](https://github.com/pace-neutrons/Horace/issues/73)). 
+
+This will:
+
+- tag the `master` branch with the release number (`Rm_n_o`)
+- copy the built artifact to the GitHub
+
+The release tag will be an anchor for any subsequently needed release branch and patch branches.
+
+The version number is stored in a text file (`VERSION`) in the root of the Herbert and Horace source. This will follow [semantic versioning](https://semver.org/) and is used in the build process to set the version number in the MATLAB and compiled C++ components.
+
+### Facility Deployment
+
+Users participating in ISIS experiments use expensive neutron facility and need to be able to see and analyse their data as quickly as possible, ideally in the process of experiment and immediately after that. In addition to standard Horace scripts users sometimes write their own scripts, which places high demand on the software reliability. Sometimes, their custom scripts use standard software in the operational modes, which are difficult to predict. Often these script identify issues, which have not been anticipated during development so users need urgent fixing or modifications to the software. Sometimes, the issues can be caused by specific user's data, so can be easily reproduced only using these data. The changes fixing the user issue should occur immediately, as users should not lose a day or even an hour waiting until the whole build pipeline completes, produces all possible artefacts and get installed on the facility computers.  
+
+Quick modifications to the software with only partial testing carries the risks of breaking the other parts of the software. Despite that, in reality, single crystal experiments on different machines rarely happen simultaneously. The combined probability of error in modified code and the situation when this error immediately and severely affects another high priority user can be considered low. If this happens, the possibility to revert changes as rapidly as they were introduced, is also very important. 
+
+Because of the specific demands, the process of deploying the software on the ISIS facility differs from the  normal build-release process. 
+
+Namely:
+
+When a release is produced and placed to the github, the master branch is tagged with the release number. 
+The release tagged version is checked out in the common software area of the ISIS computing facility provided to users. The low level build artefacts i.e. C++ libraries are copied into appropriate places of this area from the Jenkins node, testing release on this system. 
```

Comment:
Having the source on iDAaaS make me a bit uncomfortable. One of our main release platforms will never use a release generated by our release pipelines?

If what we need is a fast way to distribute a particular Git branch to users (i.e. without running the > hour long tests). There's a better way of achieving this:

We can create a new "hot-fix" pipeline, that does not run tests. Only builds and packages.

- Open a bug-fixing branch and implement your fix
- Run your branch through a "hot-fix" pipeline that just does build and package
	- This pipeline should take ~5 mins
- Retrieve the build artifact and install that on iDAaaS
- Open PR to merge your bug-fixing branch into the release branch
	- This will run all the tests - letting us know if we've broken something in ~ 1hr
	- This can be merged into the release branch
- We replace the hot-fix package with the release build artifact as soon as possible

The advantages of this are that:
- we (most of the time) have a tested and built version of Horace on iDAaaS
- we can rapidly update the version with a well-defined build
- any hot-fixed versions will have the commit-SHA appended to the version, so we get    traceability 
- we are no longer required to maintain a separate production branch
	- the only "maintained" branches are the release branches and master

Reply:
This pipeline would, obviously, require that the developer has added new tests that are able to reproduce the bug (before fix) and so demonstrate that it is has been resolved when run on the developers machine (not on the iDAaaS deployment platform) after the fix is applied and “reasonable” execution of other test cases has been performed.

Reply:
iDaaaS users indeed do not have access to the packaged release unless packaged release is equal to the repository version. In addition, currently packaged release is p-coded on the private functions, which creates problems for bugfixing. I am slowly disabling this feature by-steals but never had explicit approvement for that. c-make pipeline fully disables this option, which it probably should. 

creating tests which reproduce bug may not be easy. This should be a purpose, but its not practical in all circumstances. 

Harry's pipeline looks attractive until you look at the way how real problem is discovered. 
Thanks god, we have a standard user tests so standard setup and workflow would work ok, but let's assume that something happened and a user reports error and you are invited to help him to process his data. 

1) you got users screen and enter user enviroment which may be complex (sometimes very complex e.g. Radu)
2) you look at the errors in user setup and its 50% chance that you fix the error on this stage. good. 
3) You are unlucky --it looks like user did everything right but Matlab code returns bunch of erros. 
4) at this stage error may be in user data, user setup or our code.  According to Harry, at this stage I need to go to my machine and start trying to recriate the user issue. 
5) Recriating thing which look like user setup are not easy. I would not spend half of a day to recriate Radu's setup especially if his experiment is running and after this time my setup will work. I need to check what is going on with user setup so open the debugger and go into the code. 
6) 2/3 it something with the data or user setup messing with our code. Problem fixed or moved to other place (data are corrupted)
7) with current ratio 1/3 is the bug in the code or something which looks like a bug in our code. 
8) Here is the judgement moment. Can I fix it without affecting other users? If I am unsure, I am telling user "sorry" and follow Harrys pipeline starting from recriating user setup and data on my enviroment. Very often, at this stage I can suggest user a workaround so we continue to work indepenently. 
9) Looks like simple bug fixing which not affect other users. I am fixing it, run short sequence of tests, probably related to the bug, and commit the change. User continues the experiment and I am creating PR to fully test the fixture and incorporate it into master.  all OK. 
10) I am not sure that the fix is not affecting other users but user is in hurry and it looks like 
safe way to go. There is normally low probablility that other users are using the same code at this particular moment of time so I do a fix. User works and I am following 8-9.
11) oops -- something went wrong at 9-10. Other users start complaining. (happened once or twice -- not that often for me to remember). 
there is high probability that the first user did everything he need at this stage. Noramlly at this stage I have open channel of communication with this user so see what is going on. Anyway, I am reverting the change I did (checkout branch), discuss issue with the first user, got some solution for him (including swithching him to development channel) and continue from point - 8. People not very happy, but see that their problems are dealt with so they are OK

Harrys pipeline constrains me to point 5 of the workflow and deprives me of 2/3 probablity of fixing issues without doing much work. with ~0.1 probabilty to create additional problems which are also fixable without big problem. 

Making correct and recriatable bug report is not an easy job which no one user likes to do and not everybody even able to produce. This link in Mantid is broken and I am extreamly happy that Horace is different.  Mantid is better now, (after 10 years of development) but the number of issues and unhappy users was huge on the way and some are still there.  As we have unique advantage (the core of the code is written in high level language) let's not loose it. 

I am completely against breaking working system for the sake of an abstract idea.